### PR TITLE
View attachments #1064

### LINKED ISSUE #1064: View Attachments
#### Tasks:
- [ ] Create a **TableView** to display all attachments associated with a given `entity_id`.
- [ ] Use **Material React Table (MRT)** for the implementation.
- [ ] Ensure the table columns include:
  - **Filename**
  - **Title**
  - **Description**
  - **Created Time**
  - **Modified Time**
- [ ] Implement **pagination** and ensure it matches the pagination style used in other tables within the Inventory Management System (IMS).
- [ ] Generate mock data (chatgpt)

### LINKED ISSUE #1127: Enhance attachments table view filtering

### PR DESCRIPTION
**Pull request by rowan04**
## Description

- Created a TableView to display all attachments associated with a given entity_id, using Material React Table (MRT) for the implementation.
- The table columns include: File name, Title, Description, Created Time, Modified Time
- Implements pagination that matches the pagination style used in other tables within the Inventory Management System (IMS).
- Created attachment mock data, and get attachments handlers and methods, and tests for them.
- Adds unit and cypress tests for the attachment table, and a snapshot file
- Adds attachments table view filtering, as part of #1127

## Testing instructions

Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile board tracking

Resolves #1064
Resolves #1127

### CODE REVIEW

File: src/api/attachments.tsx [hunk 1]
```diff
@@ -24,3 +30,43 @@ export const usePostAttachmentMetadata = (): UseMutationResult<
...
+  });
+};
+
+const getAttachments = async (
+  entityId: string
+): Promise<AttachmentPostMetadataResponse[]> => {
+  const queryParams = new URLSearchParams();
+  queryParams.append('entity_id', entityId);
+  queryParams.append('modified_time', '2024-01-02T13:10:10.000+00:00')
+  queryParams.append('created_time', '2024-01-01T12:00:00.000+00:00')
```
Comment (joelvdavies) [inline src/api/attachments.tsx:55]: These shouldn't be added as the backend doesn't have these filters (at least yet).

File: src/api/attachments.tsx [hunk 2]
```diff
@@ -24,3 +30,43 @@ export const usePostAttachmentMetadata = (): UseMutationResult<
...
+      params: queryParams,
+    })
+    .then((response) => response.data);
+};
+
+export const useGetAttachments = (
+  entityId?: string
+): UseQueryResult<AttachmentPostMetadataResponse[], AxiosError> => {
+  return useQuery({
+    queryKey: ['Attachments', entityId],
```
Comment (joelvdavies) [inline src/api/attachments.tsx:68]: I think the usePostAttachmentMetadata should have something like
```typescript
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: ['Attachments'],
      });
    },
```
To ensure this will re-fetch (I am not quite sure what the caching behaviour is at this point so may not effect anything now but could if we add caching back in - the older queries have this still).

File: src/api/attachments.tsx (hunk 1 above)
Comment (rowan04) [reply src/api/attachments.tsx:55]: yeah, now I made that change in object storage API, I can write `created_time` and `modified_time` straight into `mocks/attachment.json` I think.

File: src/common/attachments/attachmentsTable.component.tsx
[the remark this answers is not in the document]
Comment (rowan04) [reply src/common/attachments/attachmentsTable.component.tsx:1]: i implemented the way the table height is calculated in the systems table. I was a bit unsure what to change.

File: src/api/attachments.tsx (hunk 2 above)
Comment (rowan04) [reply src/api/attachments.tsx:73]: Done what i think you mean.

Comment (joelvdavies) [reply src/api/attachments.tsx:73]: Yes and that fixed it so that it now updates the table when a new attachment is uploaded as its invalidating the query causing it to fetch the updated list.

File: src/common/attachments/attachmentsTable.component.tsx
[the remark this answers is not in the document]
Comment (joelvdavies) [reply src/common/attachments/attachmentsTable.component.tsx:1]: They are the same size now so it worked.

**Comment by rowan04** [comment]
I don't know why this test keeps failing, the cypress tests all pass for me locally.

**Comment by joshdimanteto** [comment]
> I don't know why this test keeps failing, the cypress tests all pass for me locally.

Is that in both interactive mode (with the GUI, i.e., `yarn e2e:interactive`) and headless mode (in the terminal, i.e., `yarn e2e`) locally?

**Comment by rowan04** [comment]
> > I don't know why this test keeps failing, the cypress tests all pass for me locally.
> 
> Is that in both interactive mode (with the GUI, i.e., `yarn e2e:interactive`) and headless mode (in the terminal, i.e., `yarn e2e`) locally?

No, I haven't tried 'headless' mode, I'm not sure I've ever tried it.

**Comment by rowan04** [comment]
> Test to see if it still fails in headless mode locally. The tests should output screenshots here `cypress/screenshots` if it fails

Fixed - I don't know why but the scrollIntoView wasn't working, so I had to manually tell it to scroll to the right.

**Comment by joshdimanteto** [comment]
> > Test to see if it still fails in headless mode locally. The tests should output screenshots here `cypress/screenshots` if it fails
> 
> Fixed - I don't know why but the scrollIntoView wasn't working, so I had to manually tell it to scroll to the right.

https://github.com/ral-facilities/inventory-management-system/pull/1270

This pr it should fix he issue and you should be able to use 

```
     it('should render in table headers', () => {
      cy.findByText('5YUQDDjKpz2z').click();
      cy.findByText(
        'High-resolution cameras for beam characterization. 1'
      ).should('exist');

      cy.findByText('Attachments').click();
      cy.findByText('Filename').should('be.visible');
      cy.findByText('Title').should('be.visible');
      cy.findByText('Description').should('be.visible');
      cy.findByText('Last modified').scrollIntoView();
      cy.findByText('Created').should('be.visible');
      cy.findByText('Last modified').should('be.visible');
    });
```

File: cypress/e2e/with_mock_data/items.cy.ts [hunk 3]
```diff
@@ -636,6 +636,70 @@ describe('Items', () => {
...
+
+      cy.findByText('Attachments').click();
+      cy.findByText('Filename').should('be.visible');
+      cy.findByText('Title').should('be.visible');
+      cy.findByText('Description').should('be.visible');
+      // Scroll 750 pixels to the right, so created and last modifed headers are visible.
+      // scrollIntoView() wasn't working in headless mode.
+      cy.scrollTo(750, 0)
+      cy.findByText('Created').should('be.visible');
+      cy.findByText('Last modified').should('be.visible');
```
Comment (joshdimanteto) [inline cypress/e2e/with_mock_data/items.cy.ts:654]:
```suggestion
      cy.findByText('Last modified').scrollIntoView();
      cy.findByText('Created').should('be.visible');
      cy.findByText('Last modified').should('be.visible');
```

Change this once #1270 is merged into develop

File: src/api/attachments.tsx [hunk 4]
```diff
@@ -19,8 +26,52 @@ export const usePostAttachmentMetadata = (): UseMutationResult<
...
+  const queryClient = useQueryClient();
   return useMutation({
     mutationFn: (attachmentMetadata: AttachmentPostMetadata) =>
       postAttachmentMetadata(attachmentMetadata),
+    onSuccess: () => {
+      queryClient.invalidateQueries({
+        queryKey: ['Attachments'],
+      });
+    },
+  });
```
Comment (joshdimanteto) [inline src/api/attachments.tsx:38]: Currently, this invalidates all `Attachments` queries after a successful post request.  

This isn't ideal for our use case because we use Uppy, which allows multiple simultaneous uploads. Invalidating all queries every time an upload completes is excessive.  

Instead, we should only invalidate the query for the attachments related to the current entity (the page we are on). There's an example of this in `UploadImagesDialog`.

File: src/api/attachments.tsx [hunk 5]
```diff
@@ -19,8 +26,52 @@ export const usePostAttachmentMetadata = (): UseMutationResult<
...
+};
+
+export const useGetAttachments = (
+  entityId?: string
+): UseQueryResult<AttachmentPostMetadataResponse[], AxiosError> => {
+  return useQuery({
+    queryKey: ['Attachments', entityId],
+    queryFn: () => getAttachments(entityId ?? ''),
+    enabled: !!entityId,
   });
```
Comment (joshdimanteto) [inline src/api/attachments.tsx:76]:
```suggestion
export const useGetAttachments = (
  entityId: string
): UseQueryResult<AttachmentPostMetadataResponse[], AxiosError> => {
  return useQuery({
    queryKey: ['Attachments', entityId],
    queryFn: () => getAttachments(entityId),
  });
```

If there is not a valid entityId the page would not load so it doesn't need to be optional

File: src/manufacturer/manufacturerTable.component.test.tsx [hunk 6]
```diff
@@ -132,5 +132,7 @@ describe('Manufacturer Table', () => {
     await waitFor(() => {
       expect(screen.getByText('Manufacturer A')).toBeInTheDocument();
     });
+
+    expect(clearFiltersButton).toBeDisabled();
```
Comment (joshdimanteto) [inline src/manufacturer/manufacturerTable.component.test.tsx:136]: This assertion is not need in this test. This test used the table row values for the assertion.

**Review by joshdimanteto — CHANGES_REQUESTED** [review]
Works really well, only some minor changes

**Review by asuresh-code — CHANGES_REQUESTED** [review]
Looks good and works well overall, just have 1 minor suggestion

File: src/mocks/Attachments.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        "id": "1",
+        "modified_time": "2024-01-02T13:10:10.000+00:00",
+        "created_time": "2024-01-01T12:00:00.000+00:00"
+    },
+
+    {
+        "entity_id": "1",
+        "file_name": "camera-setup-guide.docx",
+        "title": "camera-setup-guide",
+        "description": "Guide providing instructions for setting up and configuring cameras used in experiments.",
```
Comment (asuresh-code) [inline src/mocks/Attachments.json:26]:
```suggestion
        "entity_id": "1",
        "file_name": "camera-setup-guide.docx",
        "title": "Camera Setup Guide",
        "description": "Guide providing instructions for setting up and configuring cameras used in experiments.",
```
I think it's more appropriate for the title to be formatted without the hyphens.

File: src/common/attachments/attachmentsTable.component.tsx [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+export interface AttachmentTableProps {
+  entityId?: string;
+}
+
+function AttachmentsTable(props: AttachmentTableProps) {
+  const { entityId } = props;
+  const { data: attachments, isLoading: attachmentIsLoading } = useGetAttachments(entityId);
+
+  const columns = React.useMemo<MRT_ColumnDef<AttachmentPostMetadataResponse>[]>(() => {
```
Comment (joshdimanteto) [inline src/common/attachments/attachmentsTable.component.tsx:38]:
```suggestion
  const columns = React.useMemo<MRT_ColumnDef<AttachmentPostMetadataResponse>[]>(() => {
```

This is not the correct type for the attachment list component. Look at backend models / schemas

**Review by joshdimanteto — CHANGES_REQUESTED** [review]

File: src/common/attachments/attachmentsTable.component.tsx (hunk 8 above)
Comment (rowan04) [reply src/common/attachments/attachmentsTable.component.tsx:38]: I had tried using `AttachmentPostMetadata` originally but ran into problems with it (I can't remember what now). The reason I used `AttachmentPostMetadataResponse` is it has the `id` property (https://github.com/ral-facilities/inventory-management-system/blob/develop/src/api/api.types.tsx), like `APIImage` does which is used by the image gallery component columns.

Comment (joshdimanteto) [reply src/common/attachments/attachmentsTable.component.tsx:38]: You need to create a new type that matches the backend type

**Review by joshdimanteto — CHANGES_REQUESTED** [review]

File: src/common/attachments/attachmentsTable.component.tsx [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+  TableCellOverFlowTipProps,
+  TableHeaderOverflowTip,
+  displayTableRowCountText,
+  formatDateTimeStrings,
+  getInitialColumnFilterFnState,
+} from '../../utils';
+import { usePreservedTableState } from '../preservedTableState.component';
+
+export interface AttachmentTableProps {
+  entityId?: string;
```
[the remark this answers is not in the document]
Comment (rowan04) [reply src/common/attachments/attachmentsTable.component.tsx:31]: I agree but vscode was flagging it as an error

File: src/api/attachments.tsx (hunk 4 above)
Comment (joshdimanteto) [reply src/api/attachments.tsx]: @rowan04

File: src/api/api.types.tsx [hunk 10]
```diff
@@ -220,6 +220,12 @@ export interface Item
 }
 // ------------------------------------ OBJECT STORAGE API -----------------------------------------
 
+export interface ObjectFilePatch {
+  file_name?: string;
+  title?: string | null;
+  description?: string | null;
+}
+
```
Comment (asuresh-code) [inline src/api/api.types.tsx:228]: [This PR](https://github.com/ral-facilities/inventory-management-system/pull/1267) also changes the same api types. Could copy that implementation here or sort it out when merging, not sure what @joshdimanteto thinks?

File: src/api/attachments.test.tsx [hunk 11]
```diff
@@ -46,4 +46,18 @@ describe('attachments api functions', () => {
       });
     });
   });
+
+  describe('useGetAttachments', () => {
```
Comment (asuresh-code) [inline src/api/attachments.test.tsx:50]: Like images, I think attachments should also have a test for a successful empty list response. In the handler a value of `90` for the `entity_id` returns an empty list which you can use.

**Review by asuresh-code — CHANGES_REQUESTED** [review]
Works well, just found some more minor changes.

File: src/api/api.types.tsx (hunk 10 above)
Comment (joshdimanteto) [reply src/api/api.types.tsx:228]: That change doesn’t need to be included in this PR; it just depends on which PR gets merged into develop first.

File: src/mocks/handlers.ts [hunk 12]
```diff
@@ -1000,18 +1001,63 @@ export const handlers = [
...
+          };
+        } else if (Number(id) % 4 === 3) {
+          attachment = AttachmentsJSON[3];
+        }
+        return {
+          ...attachment,
+          id: String(id),
+        };
+      });
+    };
```
[the remark this answers is not in the document]
Comment (rowan04) [reply src/mocks/handlers.ts:1056]: yep, good spot, was unintentional

[the remark this answers is not in the document]
Comment (joshdimanteto) [reply src/mocks/handlers.ts:1056]: Can you just remove this code (get attachment handler ). As this is meant for the download pr and cannot be tested properly in this pr

[the remark this answers is not in the document]
Comment (rowan04) [reply src/mocks/handlers.ts:1056]: this code is part of the get attachments handler, I have already removed the get attachment handler from this PR.

File: src/api/attachments.test.tsx (hunk 11 above)
Comment (joshdimanteto) [reply src/api/attachments.test.tsx:50]: That test was specifically for primary images, not for fetching images using just an entity_id. This case is more similar to this PR. Additionally, the empty list scenario is already covered in the table view tests.

Comment (rowan04) [reply src/api/attachments.test.tsx:50]: so, the empty list response test for attachments doesn't need adding?

Comment (joshdimanteto) [reply src/api/attachments.test.tsx:50]: yes

File: src/mocks/handlers.ts [hunk 13]
```diff
@@ -1000,18 +1001,63 @@ export const handlers = [
...
+
+    return HttpResponse.json(generateAttachments(), { status: 200 });
+  }),
+
```
Comment (joshdimanteto) [inline src/mocks/handlers.ts:1060]:
```suggestion
  http.get('/attachments', ({ request }) => {
    const url = new URL(request.url);
    const attachmentParams = url.searchParams;
    const entityId = attachmentParams.get('entity_id');

    const generateAttachments = () => {
      return Array.from({ length: 20 }, (_, index) => {
        const id = index + 1;
        const attachment = { ...AttachmentsJSON[id % 4] };

        return {
          ...attachment,
          id: String(id), // Ensure the id is a string
          entity_id: entityId,
        };
      });
    };

    return HttpResponse.json(generateAttachments(), { status: 200 });
  }),
```

This handler can be implemented like this. The description being undefined was a specific condition needed for the image testing

File: src/mocks/handlers.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (joshdimanteto) [reply src/mocks/handlers.ts:1056]: look at https://github.com/ral-facilities/inventory-management-system/pull/1296#discussion_r1987581640

**Review by joshdimanteto — CHANGES_REQUESTED** [review]
Looks good just one minor comment.

File: src/common/attachments/attachmentsTable.component.tsx [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        showProgressBars: attachmentIsLoading, // or showSkeletons
+    },
+
+    // MUI
+    muiTableBodyRowProps: ({ row }) => {
+      return { component: TableRow, 'aria-label': `${row.original.file_name} row` };
+    },
+    muiTablePaperProps: { sx: { maxHeight: '100%' } },
+    muiTableContainerProps: ({ table }) => ({
+      sx: {
```
[the remark this answers is not in the document]
Comment (joshdimanteto) [reply src/common/attachments/attachmentsTable.component.tsx:155]: #1302

**Review by joshdimanteto — COMMENTED** [review]
https://github.com/ral-facilities/inventory-management-system/pull/1270

add the mrtTheme to the table

File: src/common/attachments/attachmentsTable.component.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (rowan04) [reply src/common/attachments/attachmentsTable.component.tsx:155]: as in like this? 
image

**Review by joshdimanteto — APPROVED** [review]
LGTM

**Comment by joshdimanteto** [comment]
> LGTM

all comments have been addressed. So I'm going to merge this pr
